### ISSUE TITLE
PhysicsFS 4.0 plans...

### ISSUE THREAD

**Issue by icculus** [icculus/physfs#41]
This is preliminary, but here's some initial wishlist things I'd like to do:

- Make the Android stuff not use PHYSFS_Init with a magic structure. This was a wrong move.
- Move all the global state into a struct and let people create instances, each with their own search path, write dir, mutexes, etc. All the functions will take an instance pointer; for people that don't care about this or want to build legacy code, we'll leave the original entry points alone, and those will manage an internal instance (or write a physfs3-compat library).
- Remove CD-ROM support (entry point will remain but never find any discs).
- Multiple write directories; they'll interpolate and mount and such like the search path does.
- Queries for other writeable directories (config/cache/etc), instead of just a "pref path"
- Probably other things.

**Comment by Fothsid** [comment]
Built-in UDF .iso support would be neat.

**Comment by icculus** [comment]
- Some sort of way to signal that Emscripten should persist some files and not others (MEMFS vs IndexedDB), and possibly a "commit" API (needed by Emscripten and Nintendo Switch).

**Comment by Green-Sky** [comment]
> * Some sort of way to signal that Emscripten should persist some files and not others (MEMFS vs IndexedDB), and possibly a "commit" API (needed by Emscripten and Nintendo Switch).

I worked around that for emscripten by mounting IDB, and setting PhysFS write dir to the same directory. (iirc)

**Comment by vectorstorm** [comment]
- file move/rename support.

**Comment by vectorstorm** [comment]
> Multiple write directories;  they'll interpolate and mount and such like the search path does

Related to this, one current pain point for me with PhysFS has been how the write path is treated like a separate filesystem which doesn't necessarily match the structure of the main virtual filesystem.

In my game, I want my writeable files to be mounted somewhere *outside* of the root of my game's main virtual filesystem, so that I can know for certain that files written by the game at runtime cannot accidentally override files which were provided by the game;  the whole write directory gets mounted over into a side "/user/" virtual directory in the search path.

This means that my code has to always know and remember that if I write a file out to "/file.txt", that same file would need to be read from "/user/file.txt".  This *does* work in PhysFS right now as long as I'm diligent about it (and I've written a wrapper around PhysFS to try to hide it from game code), but it's obviously a bit of a cognitive load that feels like maybe PhysFS could simplify.


Proposal:  maybe this whole thing could be implemented by having `PHYSFS_mount()` include some configuration flags that say for each mount point whether that path can be written to;  maybe turn its current `append?1:0` parameter into a bitfield with named parameters that can be bitwise-or'd together to denote which mounting points should allow writes.  Maybe something like this:

```cpp
#define PHYSFS_READ (0) // implicit
#define PHYSFS_LOWPRIORITY (0) // implicit if HIGHPRIORITY isn't set.
#define PHYSFS_HIGHPRIORITY (1<<0) // matches "1 == append to search path" current functionality
#define PHYSFS_WRITE (1<<1)
#define PHYSFS_READWRITE (PHYSFS_READ|PHYSFS_WRITE)

PHYSFS_init(argv[0]);
PHYSFS_mount("MyZippedData.zip", "/data", PHYSFS_HIGHPRIORITY );
PHYSFS_mount( SDL_GetPrefPath(companyName, title), "/preferences", PHYSFS_READWRITE | PHYSFS_LOWPRIORITY);
PHYSFS_mount( SDL_GetSaveGamePath(companyName, title), "/saves", PHYSFS_READWRITE | PHYSFS_LOWPRIORITY);
  // SDL doesn't actually currently provide a 'SDL_GetSaveGamePath()' API, but just as an example.

PHYSFS_File *file;
file = PHYSFS_openWrite("myfile.txt"); // fails, not in a READWRITE directory.
file = PHYSFS_openWrite("/preferences/myfile.txt"); // succeeds
// (...)
PHYSFS_close(file);
file = PHYSFS_openRead("/preferences/myfile.txt"); // also succeeds, opens the same file saved above.
file = PHYSFS_openWrite("/data/myfile.txt"); // fails, as there's no 'WRITE' path mapped here. 
file = PHYSFS_openWrite("/preferences/../myfile.txt"); // fails, as this resolves into a filepath which doesn't have a anything mounted.
```

**Comment by kleszcz** [comment]
> * Remove CD-ROM support (entry point will remain but never find any discs).

Does it mean that physical media for Xbox Series X and PS5 won't be handled?

**Comment by icculus** [comment]
> Does it mean that physical media for Xbox Series X and PS5 won't be handled?

Mmm...that's a good question. Maybe we'll leave this API after all. Let me think on it.

**Comment by icculus** [comment]
> In my game, I want my writeable files to be mounted somewhere outside of the root of my game's main virtual filesystem, so that I can know for certain that files written by the game at runtime cannot accidentally override files which were provided by the game; I mount the whole write directory over into a side "/user/" virtual directory in the search path's virtual filesystem.

I'm probably misunderstanding the problem, and if so I apologize, but why don't you mount the write dir's path last, to "/", and put it at the end of the search path, so it never overrides _anything_ and you don't have to manage it as a separate /user path when reading?

(Or maybe, once I build instances, you have a second instance that only has the write path mounted as "/", so if PHYSFS_openRead() fails, you can just try the second instance, so it's extremely isolated against accidental access. Maybe?)

**Comment by NebularNerd** [comment]
Is #18 under consideration? As mentioned if you have a file inside a zip then change it (for example a default config file), the change is wrote to the overlay. However the change is not seen and only the original file is picked up, ideally the overlay version should override the .zip original.

Currently the only solution is to see what file is changed then delete it from inside the .zip to then let the new one be seen.

**Comment by icculus** [comment]
PhysicsFS 4.0 Wishlist item:

- Remove dependency on the C runtime, the same way we did for SDL.

I think for the most part it's just str* and mem* functions. Probably won't slot dlmalloc in here, but make it so you can build PhysicsFS without a reference to malloc() where the app _must_ specify their own allocator at startup.

**Comment by icculus** [comment]
- Change coding style to be less 1990's Ryan and more 2020's Ryan.  :)

**Comment by iryont** [comment]
> I wish the issue #13 would be fixed for PhysFS 4.0. Because I've tried before to use PhysFS to load multiple assets in parallel, and then found out it didn't speed up things (actually it made loading slower), details of why that happens is mentioned in the blog post of the issue.

I came here to actually say the same thing and then I read your post.

Anyway, I believe this is the upmost important thing, even for current PhysicsFS 3.0.

**Comment by icculus** [comment]
> Anyway, I believe this is the upmost important thing, even for current PhysicsFS 3.0.

I agree, although the most immediate workaround is to not use the callback enumeration mechanism, so you end up with a complete list of files before starting to process data, and then it won't block on the mutex, and that can be done by the app right now.

I suspect the correct fix inside PhysicsFS, for 3.x if not in general, is to have each archiver promise thread safety directly, which might mean duplicating the archive file handle in the same way we do when opening individual files within it, so it can seek around to enumerate without interfering with other operations that might occur, but I haven't had a moment to untangle the whole mess yet.

**Comment by iryont** [comment]
> > Anyway, I believe this is the upmost important thing, even for current PhysicsFS 3.0.
> 
> I agree, although the most immediate workaround is to not use the callback enumeration mechanism, so you end up with a complete list of files before starting to process data, and then it won't block on the mutex, and that can be done by the app right now.
> 
> I suspect the correct fix inside PhysicsFS, for 3.x if not in general, is to have each archiver promise thread safety directly, which might mean duplicating the archive file handle in the same way we do when opening individual files within it, so it can seek around to enumerate without interfering with other operations that might occur, but I haven't had a moment to untangle the whole mess yet.

I suppose the main problem is reading files from the same archive (handle) in multiple threads (e.g. assets.zip). Just like you said, having duplicated archive file handle could solve this issue once for all.

**Comment by MikuAuahDark** [comment]
I'd love if PhysFS is able to mount Android `content://` URI that's retrieved from [`ACTION_OPEN_DOCUMENT_TREE` intent](https://developer.android.com/training/data-storage/shared/documents-files#grant-access-directory) or being able to create PhysFS_IO from an Android `content://` URI.

**Comment by icculus** [comment]
Split out public header: #71

**Comment by williamjcm** [comment]
I personally would like improved documentation, especially related to custom archivers.

I'm currently trying to make one for a custom format, but the lack of documentation in Doxygen, combined with the built-in archivers' source code barely having any comments, are making it harder than it should be.
